pipeline/psth.py

```diff
@@ -19,129 +22,229 @@
+    ---
+    trial_condition_desc:       varchar(1000)   # trial condition description
+    trial_condition_func:       varchar(36)     # trial retrieval function
+    trial_condition_arg:        longblob        # trial retrieval arguments
+    """
 
+    @property
+    def contents(self):
+        contents_data = (  # FIXME: functions should handle missing args
+            {
+                'trial_condition_desc': 'good_noearlylick_hit',
+                'trial_condition_func': '_get_trials_no_stim',
+                'trial_condition_arg': {
+                    'task': 'audio delay',
+                    'task_protocol': 1,
+                    'outcome': 'hit',
+                    'early_lick': 'no early'}
+            },
+            {
+                'trial_condition_desc': 'good_noearlylick_left_hit',
+                'trial_condition_func': '_get_trials_no_stim',
+                'trial_condition_arg': {
+                    'task': 'audio delay',
+                    'task_protocol': 1,
+                    'outcome': 'hit',
+                    'early_lick': 'no early',
+                    'trial_instruction': 'left'}
+            },
+            {
+                'trial_condition_desc': 'good_noearlylick_right_hit',
+                'trial_condition_func': '_get_trials_no_stim',
+                'trial_condition_arg': {
+                    'task': 'audio delay',
+                    'task_protocol': 1,
+                    'outcome': 'hit',
+                    'early_lick': 'no early',
+                    'trial_instruction': 'right'}
+            },
+            {
+                'trial_condition_desc': 'good_noearlylick_left_miss',
+                'trial_condition_func': '_get_trials_no_stim',
+                'trial_condition_arg': {
+                    'task': 'audio delay',
+                    'task_protocol': 1,
+                    'outcome': 'miss',
+                    'early_lick': 'no early',
+                    'trial_instruction': 'left'}
+            },
+            {
+                'trial_condition_desc': 'good_noearlylick_right_miss',
+                'trial_condition_func': '_get_trials_no_stim',
+                'trial_condition_arg': {
+                    'task': 'audio delay',
+                    'task_protocol': 1,
+                    'outcome': 'miss',
+                    'early_lick': 'no early',
+                    'trial_instruction': 'right'}
+            },
+        )
+
+        return ({**d, 'trial_condition_id': key_hash(d['trial_condition_arg'])}
+                for d in contents_data)
+
+    @classmethod
+    def get_trials(cls, trial_condition_desc):
+        return cls.get_func({'trial_condition_desc': trial_condition_desc})()
+
+    @classmethod
+    def get_func(cls, key):
+        self = cls()
+
+        func, args = (self & key).fetch1(
+            'trial_condition_func', 'trial_condition_arg')
+
+        return partial(dict(getmembers(cls))[func], **args)
```

Comment:
you can write this as `getattr`

Reply:
This didn't seem to 'do' for datajoint classes (seems like dj metaclass problem) e.g. given:

```
class Foo:
    def foo():
        pass
```
vs
`pipeline.UnitPsth`:


```
>>> psth.UnitPsth
<class 'pipeline.psth.UnitPsth'>
>>> Foo
<class 'pipeline.shell.Foo'>
>>> getattr(Foo, 'foo')
<function Foo.foo at 0x12ab3ec80>
>>> getattr(psth.UnitPsth, 'get_func')
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/Users/chris/src/djc/datajoint-python/datajoint/user_tables.py", line 45, in __getattribute__
    else super().__getattribute__(name))
AttributeError: type object 'UnitPsth' has no attribute 'get_func'
>>> getattr(psth.UnitPsth(), 'get_func')
Traceback (most recent call last):
  File "<console>", line 1, in <module>
AttributeError: 'UnitPsth' object has no attribute 'get_func'
```

Reply:
I see - that's weird and to be assessed but in the future.